metagraph/src/annotation/taxonomy/taxonomic_db.cpp

```diff
@@ -0,0 +1,326 @@
+    if (taxids.empty()) {
+        logger->error("Can't find LCA for an empty set of taxids.");
+        std::exit(1);
+    }
+    uint64_t left_idx = node_to_linearization_idx[taxids[0]];
+    uint64_t right_idx = node_to_linearization_idx[taxids[0]];
+    for (const auto &taxid: taxids) {
+        if (node_to_linearization_idx[taxid] < left_idx) {
+            left_idx = node_to_linearization_idx[taxid];
+        }
+        if (node_to_linearization_idx[taxid] > right_idx) {
+            right_idx = node_to_linearization_idx[taxid];
+        }
+    }
+
+    uint64_t log_dist = precalc_log2[right_idx - left_idx];
+    uint64_t left_lca = rmq_data[log_dist][left_idx];
+    uint64_t right_lca = rmq_data[log_dist][right_idx - precalc_pow2[log_dist] + 1];
+
+    if (node_depth[left_lca] > node_depth[right_lca]) {
+        return left_lca;
+    }
+    return right_lca;
+}
+
+bool TaxonomyDB::get_normalized_taxid(const std::string accession_version, NormalizedTaxId &taxid) const {
+    num_get_taxid_calls += 1;
+    if (! lookup_table.count(accession_version)) {
+        // accession_version not in the lookup_table
+        num_get_taxid_calls_failed += 1;
+        return false;
+    }
+    if (! normalized_taxid.count(lookup_table.at(accession_version))) {
+        // taxid (corresponding to the current accession_version) not in the taxonomic tree.
+        return false;
+    }
+
+    taxid = normalized_taxid.at(lookup_table.at(accession_version));
+    return true;
+}
+
+void TaxonomyDB::kmer_to_taxid_map_update(const annot::MultiLabelEncoded<std::string> &annot,
+                                          std::mutex &taxo_mutex) {
+    for (const auto &label: annot.get_all_labels()) {
+        AccessionVersion accession_version = get_accession_version_from_label(label);
+        uint64_t taxid;
+        if(!get_normalized_taxid(accession_version, taxid)) {
+            continue;
+        }
+        annot.call_objects(label, [&](const auto &index) {
+          if (index <= 0) {
+              return;
+          }
+          if (index >= taxonomic_map.size()) {
+              // Double the size of taxonomic_map until the current 'index' fits inside.
+              taxo_mutex.lock();
+              uint64_t new_size = taxonomic_map.size();
+              while (index >= new_size) {
+                  new_size = new_size * 2 + 1;
+              }
+              sdsl::int_vector<> aux_taxonomic_map = taxonomic_map;
+              taxonomic_map.resize(new_size);
+              for(uint64_t i = 0; i < taxonomic_map.size(); ++i) {
+                  taxonomic_map[i] = 0;
+              }
+              for(uint64_t i = 0; i < aux_taxonomic_map.size(); ++i) {
+                  taxonomic_map[i] = aux_taxonomic_map[i];
+              }
+              taxo_mutex.unlock();
+          }
+
+          if (taxonomic_map[index] == 0) {
+              taxo_mutex.lock();
+              taxonomic_map[index] = taxid;
+              taxo_mutex.unlock();
```

Comment:
Use lock_guard instead of doing lock() and unlock on mutexes. Check why it's safer.
```suggestion
              std::lock_guard<std::mutex> lock(taxo_mutex);
              taxonomic_map[index] = taxid;
```